### ISSUE TITLE
NavMeshLink's OffMeshLinkData.offMeshLink/OffMeshLinkData.offMeshLink.area is null.

### ISSUE THREAD

**Issue by jangofett890** [Unity-Technologies/NavMeshComponents#71]
I just determined that I need the generation of NavMesh at runtime so I downloaded this and begun converting everything to use the new system. However I ran into a problem with the new NavMeshLink.

Currently for my navigation system I have to access the area type of the Off Mesh Link to determine which action to take to cross the link (Climbing a ladder, jumping, etc). To do this I ask the agent for the next Off Mesh Link's data and this returns fine with the new NavMesh Link. However when OffMeshLinkData.offMeshLink.area is accessed via OffMeshLinkData I get an immediate nullreferenceexception "Object Reference is not set to an instance of an object." Through debugging I found it was the area field exactly, not offMeshLink. 

Here is my code:
               

                NavMeshPath path = new NavMeshPath();
                agent.CalculatePath(target.transform.position, path);
                agent.SetDestination(target.transform.position);             
                OffMeshLinkData offData = agent.nextOffMeshLinkData;

                if (Vector3.Distance(gameObject.transform.position, offData.startPos) <= reach / 2) {
                    Debug.Log(offData.linkType);
                    switch (offData.linkType) {
                        case (OffMeshLinkType.LinkTypeManual):
                                //This is where the error is occuring 3 is my leaping nav layer.
                                if (offData.offMeshLink.area == 3) {
                                    //JumpForwardCodeHere
                                } else {
                                    //TODO Ladders, etc
                                }
                           }
                            break;
                        case (OffMeshLinkType.LinkTypeDropDown):
                            //WalkForwardCodeHere
                            break;
                        case (OffMeshLinkType.LinkTypeJumpAcross):
                            //JumpForwardCodeHere
                            break;
                        default:
                            break;
                    }

**Comment by Candescence** [comment]
Yeah, this is a pretty huge bug, and I'm surprised it hasn't been fixed yet, because it makes NavMeshLinks significantly less useful if you want to use areas for different types of link navigation.

**Comment by vargalorand** [comment]
I am working on similar features right now, where I also rely on detecting the type of mesh link, and performing the necessary actions accordingly (climbing ladders, vaulting over obstacles, opening doors, etc). Unfortunately I have to stick to the "vanilla" OffMeshLinks until this issue fixed. Otherwise I would have to make unnecessarily complicated workarounds with either raycasting, collisions/triggers or something else. Hope this gets fixed soon.

**Comment by vargalorand** [comment]
I did a bit more digging between these issues (which I probably should have done before posting my previous comment...), but his issue seems to be solved by #34. I am able to correctly detect the link type now.

**Comment by jangofett890** [comment]
That issue is referencing the agent on the offmeshlinkdata not the actual link itself.

**Comment by lassade** [comment]
I found the awnser [here](https://issuetracker.unity3d.com/issues/navmesh-currentoffmeshlinkdata-dot-offmeshlink-always-returns-null-even-when-agent-is-on-offmeshlink)

TL;DR

You should watch until the ```agent.isOnOffMeshLink == true``` then you can get the ```NavMeshLink``` using ```var link = (UnityEngine.AI.NavMeshLink)agent.navMeshOwner; ```.

**Comment by vadermemo** [comment]
> I am working on similar features right now, where I also rely on detecting the type of mesh link, and performing the necessary actions accordingly (climbing ladders, vaulting over obstacles, opening doors, etc). Unfortunately I have to stick to the "vanilla" OffMeshLinks until this issue fixed. Otherwise I would have to make unnecessarily complicated workarounds with either raycasting, collisions/triggers or something else. Hope this gets fixed soon.

Hi friend, I wonder which is the status from these features from exactly one year after.

**Comment by smasson99** [comment]
var link = (NavMeshLink) navMeshAgent.navMeshOwner;

is still getting a null value for me

**Comment by lassade** [comment]
`navMeshAgent.navMeshOwner` is only valid when `agent.isOnOffMeshLink == true`: use this
```
IEnumerator FunctionA()
{
  yield return new WaitUntil(() => navMeshAgent.isOnOffMeshLink);
  var link = (UnityEngine.AI.NavMeshLink)navMeshAgent.navMeshOwner;

  // Your logic goes here ...
}
```

**Comment by lassade** [comment]
Not sure I used this code with Unity 2018, not sure whats the catch is on the newer version, maybe you have a better luck on forums

**Comment by lassade** [comment]
In this situation you may want to log the state of some variables every frame to then analize whats goning on:
```
void Update() {
  Debug.Log($"{navMeshAgent.isOnOffMeshLink}, {navMeshAgent.navMeshOwner}");
}
```

**Comment by hszhsh** [comment]
I'm using Unity 2020. The problem is when navMeshAgent.isOnOffMeshLink is true, for the most time navMeshAgent.isOnOffMeshLink will return the NavMeshLink, but sometimes navMeshAgent.isOnOffMeshLink is null
